go/tasks/plugins/array/k8s/subtask.go

```diff
@@ -0,0 +1,346 @@
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		if k8serrors.IsForbidden(err) {
+			if strings.Contains(err.Error(), "exceeded quota") {
+				logger.Warnf(ctx, "Failed to launch job, resource quota exceeded and the operation is not guarded by back-off. err: %v", err)
+				return pluginsCore.PhaseInfoWaitingForResourcesInfo(time.Now(), pluginsCore.DefaultPhaseVersion, fmt.Sprintf("Exceeded resourcequota: %s", err.Error()), nil), nil
+			}
+			return pluginsCore.PhaseInfoRetryableFailure("RuntimeFailure", err.Error(), nil), nil
+		} else if k8serrors.IsBadRequest(err) || k8serrors.IsInvalid(err) {
+			logger.Errorf(ctx, "Badly formatted resource for plugin [%s], err %s", executorName, err)
+			// return pluginsCore.DoTransition(pluginsCore.PhaseInfoFailure("BadTaskFormat", err.Error(), nil)), nil
+		} else if k8serrors.IsRequestEntityTooLargeError(err) {
+			logger.Errorf(ctx, "Badly formatted resource for plugin [%s], err %s", executorName, err)
+			return pluginsCore.PhaseInfoFailure("EntityTooLarge", err.Error(), nil), nil
+		}
+		reason := k8serrors.ReasonForError(err)
+		logger.Errorf(ctx, "Failed to launch job, system error. err: %v", err)
+		return pluginsCore.PhaseInfoUndefined, errors.Wrapf(stdErrors.ErrorCode(reason), err, "failed to create resource")
+	}
+
+	return pluginsCore.PhaseInfoQueued(time.Now(), pluginsCore.DefaultPhaseVersion, "task submitted to K8s"), nil
+}
+
+// finalizeSubtask performs operations to complete the k8s pod defined by the SubTaskExecutionContext
+// and Config. These may include removing finalizers and deleting the k8s resource.
+func finalizeSubtask(ctx context.Context, stCtx SubTaskExecutionContext, cfg *Config, kubeClient pluginsCore.KubeClient) error {
+	errs := stdErrors.ErrorCollection{}
+	var pod *v1.Pod
+	var nsName k8stypes.NamespacedName
+	k8sPluginCfg := config.GetK8sPluginConfig()
+	if k8sPluginCfg.InjectFinalizer || k8sPluginCfg.DeleteResourceOnFinalize {
+		o, err := podPlugin.DefaultPodPlugin.BuildIdentityResource(ctx, stCtx.TaskExecutionMetadata())
+		if err != nil {
+			// This will recurrent, so we will skip further finalize
+			logger.Errorf(ctx, "Failed to build the Resource with name: %v. Error: %v, when finalizing.", stCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), err)
+			return nil
+		}
+
+		pod = o.(*v1.Pod)
+
+		addMetadata(stCtx, cfg, config.GetK8sPluginConfig(), pod)
+		nsName = k8stypes.NamespacedName{Namespace: pod.GetNamespace(), Name: pod.GetName()}
+	}
+
+	// In InjectFinalizer is on, it means we may have added the finalizers when we launched this resource. Attempt to
+	// clear them to allow the object to be deleted/garbage collected. If InjectFinalizer was turned on (through config)
+	// after the resource was created, we will not find any finalizers to clear and the object may have already been
+	// deleted at this point. Therefore, account for these cases and do not consider them errors.
+	if k8sPluginCfg.InjectFinalizer {
```

Comment:
I _think_ this issue existed before... if we every enable the InjectFinalizer flag, then launch some workflows, then disable the flag, the plugin won't even attempt to clear the finalizers and these workflows will be forever stuck... am I reading this correctly?
I think we may have to either store in the task plugin state the fact we injected the finalizer or change this code to always read the resource and clear the finalizer IFF our finalizer exist...
Either way, if this is an existing issue, mind filing a github issue about it so we don't forget? if it's a new issue in this code, which way do you think we should go?

Reply:
I think you are correct. I will test to make sure.

Similar to ^^^ this came straight from the flytepropeller plugin manager [here](https://github.com/flyteorg/flytepropeller/blob/163824c650fbeaedc55a4df99a69825e8ff37306/pkg/controller/nodes/task/k8s/plugin_manager.go#L426). Left a [comment below](https://github.com/flyteorg/flyteplugins/pull/244#issuecomment-1064153003) on one option as to where this fix may be addressed. Happy to fix it anywhere.